File: 0006.zigzag-conversion.cpp
class Solution
{
public:
    string convert(string s, int numRows)
    {
        if (numRows <= 1)
        {
            return s;
        }
        vector<string> a;
        for (int i = 0; i < numRows; i++)
        {
            a.push_back("");
        }
        int now = 0;
        int delta = 1;
        for (size_t i = 0; i < s.length(); i++)
        {
            a[now] += s[i];
            if ((delta > 0 && now == numRows - 1) || (delta < 0 && now == 0))
            {
                delta *= -1;
            }
            now += delta;
        }
        string res = "";
        for (int i = 0; i < numRows; i++)
        {
            res += a[i];
        }
        return res;
    }
};

File: rust/0006_zigzag-conversion.rs
impl Solution
{
    pub fn convert(s: String, num_rows: i32) -> String
    {
        if num_rows <= 1
        {
            return s;
        }

        let nr = num_rows as usize;
        let mut res = String::new();
        let n = s.len();
        let round = nr * 2 - 2;

        for r in 0..nr
        {
            let mut i = r;

            if i == 0 || i == nr -1
            {
                while i < n
                {
                    res.push(s.as_bytes()[i] as char);
                    i += round;
                }
            }
            else
            {
                let mut step = round - r * 2;
                while i < n
                {
                    res.push(s.as_bytes()[i] as char);
                    i += step;
                    step = round - step;
                }
            }
        }

        return res;
    }
}
